File: dvrk/state_relay_delay.py
#!/usr/bin/env python

import rospy
from std_msgs.msg import String
from geometry_msgs.msg import Pose, PoseStamped
from sensor_msgs.msg import JointState

lastStatePublished = ""
lastPosePublishedPose = Pose()
lastPosePublishedPoseStamped = PoseStamped()
lastJawReceived = JointState()

state_pub = None
cartesian_pub = None
jaw_pub = None

def are_poses_equal_pose(pose1, pose2):
    return (pose1.position.x == pose2.position.x and
            pose1.position.y == pose2.position.y and
            pose1.position.z == pose2.position.z and
            pose1.orientation.x == pose2.orientation.x and
            pose1.orientation.y == pose2.orientation.y and
            pose1.orientation.z == pose2.orientation.z and
            pose1.orientation.w == pose2.orientation.w)

def are_poses_equal_pose_stamped(pose1, pose2):
    return (pose1.pose.position.x == pose2.pose.position.x and
            pose1.pose.position.y == pose2.pose.position.y and
            pose1.pose.position.z == pose2.pose.position.z and
            pose1.pose.orientation.x == pose2.pose.orientation.x and
            pose1.pose.orientation.y == pose2.pose.orientation.y and
            pose1.pose.orientation.z == pose2.pose.orientation.z and
            pose1.pose.orientation.w == pose2.pose.orientation.w)

def state_callback(msg):
    global lastStatePublished
    if msg.data != lastStatePublished:
        state_pub.publish(msg)
        lastStatePublished = msg.data
        rospy.loginfo(f"New state: {lastStatePublished}")

def cartesian_callback_pose(msg):
    global lastPosePublishedPose
    if lastStatePublished == "READY" and not are_poses_equal_pose(lastPosePublishedPose, msg):
        cartesian_pub.publish(msg)
        jaw_pub.publish(lastJawReceived)
        lastPosePublishedPose = msg

def cartesian_callback_pose_stamped(msg):
    global lastPosePublishedPoseStamped
    if lastStatePublished == "READY" and not are_poses_equal_pose_stamped(lastPosePublishedPoseStamped, msg):
        cartesian_pub.publish(msg)
        jaw_pub.publish(lastJawReceived)
        lastPosePublishedPoseStamped = msg

def jaw_callback(msg):
    global lastJawReceived
    lastJawReceived = msg

def main():
    global state_pub, cartesian_pub, jaw_pub

    rospy.init_node('state_relay', anonymous=True)

    if len(rospy.myargv()) != 8:
        rospy.logerr("Failed to launch state_relay, requires 6 arguments")
        rospy.logerr("  -- Arg 1 = subscribed state topic name")
        rospy.logerr("  -- Arg 2 = to publish state topic name")
        rospy.logerr("  -- Arg 3 = subscribed cartesian topic name")
        rospy.logerr("  -- Arg 4 = to publish cartesian topic name")
        rospy.logerr("  -- Arg 5 = subscribed jaw topic name")
        rospy.logerr("  -- Arg 6 = to publish jaw topic name")
        rospy.logerr("  -- Arg 7 = Pose or PoseStamped")
        return -1

    state_subscribe_topic = rospy.myargv()[1] # e.g. PSM1
    state_publish_topic = rospy.myargv()[2] # e.g. NetPSM1

    cart_subscribe_topic = rospy.myargv()[3]
    cart_publish_topic = rospy.myargv()[4]

    jaw_subscribe_topic = rospy.myargv()[5]
    jaw_publish_topic = rospy.myargv()[6]

    pose_or_posestamped = rospy.myargv()[7]

    state_pub = rospy.Publisher(state_publish_topic, String, queue_size=10)
    jaw_pub = rospy.Publisher(jaw_publish_topic, JointState, queue_size=10)

    if pose_or_posestamped == "Pose":
        cartesian_pub = rospy.Publisher(cart_publish_topic, Pose, queue_size=10)
    elif pose_or_posestamped == "PoseStamped":
        cartesian_pub = rospy.Publisher(cart_publish_topic, PoseStamped, queue_size=10)
    else:
        rospy.logerr("Arg 7 can only be Pose or PoseStamped!!!")
        return -1

    rospy.Subscriber(state_subscribe_topic, String, state_callback)
    
    if pose_or_posestamped == "Pose":
        rospy.Subscriber(cart_subscribe_topic, Pose, cartesian_callback_pose)
    elif pose_or_posestamped == "PoseStamped":
        rospy.Subscriber(cart_subscribe_topic, PoseStamped, cartesian_callback_pose_stamped)
    else:
        rospy.logerr("Arg 7 can only be Pose or PoseStamped!!!")
        return -1

    rospy.Subscriber(jaw_subscribe_topic, JointState, jaw_callback)

    rospy.spin()

if __name__ == '__main__':
    main()

File: dvrk/state_relay_delay.cpp
//
// Created by florian on 1/9/20.
//

#include <ros/ros.h>
#include <std_msgs/String.h>
#include <geometry_msgs/Pose.h>
#include <geometry_msgs/PoseStamped.h>
#include <sensor_msgs/JointState.h>
#include <stdio.h>
#include <queue>

std::string lastStatePublished;
geometry_msgs::Pose lastPosePublishedPose;
geometry_msgs::PoseStamped lastPosePublishedPoseStamped;
sensor_msgs::JointState lastJawRecieved;

// Queues to buffer relayed messages
std::queue<std_msgs::String::ConstPtr> state_buffer;
std::queue<geometry_msgs::PoseStamped::ConstPtr> cart_buffer;
std::queue<sensor_msgs::JointState> jaw_buffer;

ros::Publisher state_pub;
ros::Publisher cartesian_pub;
ros::Publisher jaw_pub;

bool arePosesEqualPose(geometry_msgs::Pose pose1, geometry_msgs::Pose pose2){
    if (pose1.position.x != pose2.position.x){
        return false;
    }
    if (pose1.position.y != pose2.position.y){
        return false;
    }
    if (pose1.position.z != pose2.position.z){
        return false;
    }
    if (pose1.orientation.x != pose2.orientation.x){
        return false;
    }
    if (pose1.orientation.y != pose2.orientation.y){
        return false;
    }
    if (pose1.orientation.z != pose2.orientation.z){
        return false;
    }
    if (pose1.orientation.w != pose2.orientation.w){
        return false;
    }
    return true;
}

bool arePosesEqualPoseStamped(geometry_msgs::PoseStamped pose1, geometry_msgs::PoseStamped pose2){
    if (pose1.pose.position.x != pose2.pose.position.x){
        return false;
    }
    if (pose1.pose.position.y != pose2.pose.position.y){
        return false;
    }
    if (pose1.pose.position.z != pose2.pose.position.z){
        return false;
    }
    if (pose1.pose.orientation.x != pose2.pose.orientation.x){
        return false;
    }
    if (pose1.pose.orientation.y != pose2.pose.orientation.y){
        return false;
    }
    if (pose1.pose.orientation.z != pose2.pose.orientation.z){
        return false;
    }
    if (pose1.pose.orientation.w != pose2.pose.orientation.w){
        return false;
    }
    return true;
}

void stateCallback(const std_msgs::String::ConstPtr& msg){
    if(msg->data != lastStatePublished){
        state_pub.publish(msg);
        // state_buffer.push(msg);
        lastStatePublished = msg->data;
	std::cout << "New state: " << lastStatePublished << std::endl;
    }
}

void cartesianCallbackPose(const geometry_msgs::Pose::ConstPtr& msg){
    if(lastStatePublished == "READY" && !arePosesEqualPose(lastPosePublishedPose, *msg)){
        // cartesian_pub.publish(msg);
        // jaw_pub.publish(lastJawRecieved);
		
	geometry_msgs::PoseStamped::ConstPtr new_msg; 
	new_msg->header.stamp = ros::Time::now();
	new_msg->pose.position.x = msg->position.x; 
	new_msg->pose.position.y = msg->position.y;
	new_msg->pose.position.z = msg->position.z;
	new_msg->pose.orientation.x = msg->orientation.x; 
	new_msg->pose.orientation.y = msg->orientation.y;
	new_msg->pose.orientation.z = msg->orientation.z;
	new_msg->pose.orientation.w = msg->orientation.w;
        cart_buffer.push(new_msg);
        jaw_buffer.push(lastJawRecieved);

        lastPosePublishedPose = *msg;
    }
}

void cartesianCallbackPoseStamped(const geometry_msgs::PoseStamped::ConstPtr& msg){
    if(lastStatePublished == "READY" && !arePosesEqualPoseStamped(lastPosePublishedPoseStamped, *msg)){
        // cartesian_pub.publish(msg);
        // jaw_pub.publish(lastJawRecieved);

        cart_buffer.push(msg);
        jaw_buffer.push(lastJawRecieved);

        lastPosePublishedPoseStamped = *msg;
    }
}

void jawCallback(const sensor_msgs::JointState::ConstPtr& msg){
    lastJawRecieved = *msg;
}

int main(int argc, char **argv)
{
    ros::init(argc, argv, "state_relay");
    ros::NodeHandle n;

    if(argc != 8){
        std::cerr << "Failed to launch state_relay, requires 6 arguments" << std::endl;
        std::cerr << "  -- Arg 1 = subscribed state topic name" << std::endl;
        std::cerr << "  -- Arg 2 = to publish state topic name" << std::endl;
        std::cerr << "  -- Arg 3 = subscribed cartesian topic name" << std::endl;
        std::cerr << "  -- Arg 4 = to publish cartesian topic name" << std::endl;
        std::cerr << "  -- Arg 5 = subscribed jaw topic name" << std::endl;
        std::cerr << "  -- Arg 6 = to publish jaw topic name" << std::endl;
	    std::cerr << "  -- Arg 7 = Pose or PoseStamped" << std::endl;
        return -1;
    }

    std::string state_subscribe_topic = argv[1];
    std::string state_publish_topic = argv[2];

    std::string cart_subscribe_topic = argv[3];
    std::string cart_publish_topic = argv[4];

    std::string jaw_subscribe_topic = argv[5];
    std::string jaw_publish_topic = argv[6];

    std::string pose_or_posestamped = argv[7]; 


    state_pub = n.advertise<std_msgs::String>   (state_publish_topic, 10);
    if (pose_or_posestamped == "Pose") {
	cartesian_pub = n.advertise<geometry_msgs::Pose>(cart_publish_topic,  10);
    } else if (pose_or_posestamped == "PoseStamped") {
	cartesian_pub = n.advertise<geometry_msgs::PoseStamped>(cart_publish_topic,  10);
    } else {
	std::cerr << "Arg 7 can only be Pose or PoseStamped!!!" << std::endl;
    } 
    jaw_pub = n.advertise<sensor_msgs::JointState>(jaw_publish_topic, 10);

    ros::Subscriber state_sub = n.subscribe(state_subscribe_topic, 10, stateCallback);
    ros::Subscriber cart_sub;
    if (pose_or_posestamped == "Pose") {
    	cart_sub  = n.subscribe(cart_subscribe_topic,  10, cartesianCallbackPose);
    } else if (pose_or_posestamped == "PoseStamped") {
	cart_sub  = n.subscribe(cart_subscribe_topic,  10, cartesianCallbackPoseStamped);
    } else {
	std::cerr << "Arg 7 can only be Pose or PoseStamped!!!" << std::endl;
    }
    ros::Subscriber jaw_sub = n.subscribe(jaw_subscribe_topic,   10, jawCallback);

    // ros::spin();

    // My new code goes here?
    ros::Rate loop_rate(100); // 10 Hz loop rate
    ros::Duration delay(1.0); // hardcode delay for now

    while (ros::ok()) {
        ros::spinOnce();
        ros::Time current_time = ros::Time::now();
	ros::Time front_timestamp(cart_buffer.front()->header.stamp.sec, cart_buffer.front()->header.stamp.nsec);

	// front_timestamp = cart_buffer.front()->header.stamp.sec + std::pow(cart_buffer.front()->header.stamp.nsec, -9);
        while (!cart_buffer.empty() && (current_time - front_timestamp) >= delay) {
            cartesian_pub.publish(cart_buffer.front());
            jaw_pub.publish(jaw_buffer.front());
            cart_buffer.pop();
            jaw_buffer.pop();
        }
        
        loop_rate.sleep();
    
    }
    
    return 0;
}
